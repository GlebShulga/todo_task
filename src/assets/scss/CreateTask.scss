.CreateTask {
  justify-items: center;
  padding: 1rem 1rem;
}

.CreateTask-form {
  display: flex;
  overflow: hidden;
}

.CreateTask-form_input {
  padding: 0.5rem 1rem;
  border: lightgrey;
  background-color: lightgrey;
  outline: none;

  &:focus {
    background-color: darkgrey;
  }
}

.CreateTask-form_button {
  padding: 0.5rem 1rem;
  background-color: limegreen;
  color: white;
  font-weight: 700;
  &:hover {
    background-color: lightgreen;
  }
}

.CreateTask-form_button_error {
  opacity: 0.5;
}

.Error {
  color: red;
  padding-top: 1rem;
  text-align: center;
}
