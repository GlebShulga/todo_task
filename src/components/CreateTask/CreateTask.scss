.create_task {
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
  align-items: flex-end;
  padding: 1rem 1rem;
  margin-bottom: 1rem;
}

.create_task-form {
  display: flex;
  overflow: hidden;
}

.create_task-form_input {
  padding: 0.5rem 1rem;
  border: lightgrey;
  background-color: lightgrey;
  outline: none;

  &:focus {
    background-color: darkgrey;
  }
}
