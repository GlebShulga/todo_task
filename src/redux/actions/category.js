/* eslint-disable no-loop-func */
import axios from "axios";
import {
  ADD_CATEGORY,
  GET_CATEGORIES_LIST,
  GET_CATEGORIES_TITLE_LIST,
  GET_ROOT_CATEGORY_LIST,
  RENAME_CATEGORY,
  DEL_CATEGORY,
  UPDATE_CHOSEN_CATEGORY,
  SET_IS_FILTER_STATUS_DONE,
  SET_IS_CREATE_TASK_MODAL,
} from "../types/category";

export function fetchCategoryList() {
  return (dispatch, getState) => {
    const store = getState();
    const url = store.router.location.pathname?.substring(1);
    const complexUrl = url?.match("[^/](.*?)(?=/)");
    const categoryUrl = complexUrl !== null ? complexUrl[0] : url;
    return axios
      .get("/api/v1/category")
      .then(({ data }) => {
        let current = data.find((cat) => cat.categoryTitle === categoryUrl);
        let updatedCategoryList = data;
        while (current?.parentCategoryId) {
          updatedCategoryList = updatedCategoryList.map((cat) =>
            cat.categoryId === current.parentCategoryId
              ? { ...cat, isExpanded: true }
              : cat
          );
          current = updatedCategoryList.find(
            (cat) => cat.categoryId === current.parentCategoryId
          );
        }

        const rootList = updatedCategoryList.filter(
          (el) => el.parentCategoryId === null
        );
        const titleList = data.map((category) => category.categoryTitle);
        dispatch({ type: GET_ROOT_CATEGORY_LIST, rootList });
        dispatch({
          type: GET_CATEGORIES_LIST,
          categoryList: updatedCategoryList,
        });
        dispatch({ type: GET_CATEGORIES_TITLE_LIST, titleList });
      })
      .catch((err) => console.log(err));
  };
}

export function addCategory(categoryTitle, parentCategoryId, lvl) {
  return (dispatch) => {
    axios({
      method: "post",
      url: "/api/v1/category/:categoryTitle",
      data: {
        categoryTitle,
        parentCategoryId,
        lvl,
      },
    })
      .then(({ data: categoryList }) => {
        const rootList = categoryList?.filter(
          (el) => el.parentCategoryId === null
        );
        dispatch({ type: ADD_CATEGORY, categoryList });
        dispatch({ type: GET_ROOT_CATEGORY_LIST, rootList });
      })
      .catch((err) => console.log(err));
  };
}

export function renameCategory(categoryId, categoryTitle) {
  return (dispatch) => {
    axios({
      method: "patch",
      url: "/api/v1/category",
      data: {
        categoryId,
        categoryTitle,
      },
    })
      .then(({ data: categoryList }) => {
        dispatch({ type: RENAME_CATEGORY, categoryList });
      })
      .catch((err) => console.log(err));
  };
}

export function delCategory(categoryId) {
  return (dispatch) => {
    axios({
      method: "delete",
      url: "/api/v1/category",
      data: {
        categoryId,
      },
    }).then(({ data: categoryList }) => {
      const rootList = categoryList.filter(
        (el) => el.parentCategoryId === null
      );
      dispatch({ type: DEL_CATEGORY, categoryList });
      dispatch({ type: GET_ROOT_CATEGORY_LIST, rootList });
    });
  };
}

export function updateChosenCategory(chosenCategory) {
  return { type: UPDATE_CHOSEN_CATEGORY, chosenCategory };
}

export function setIsFilterStatusDone(data) {
  return { type: SET_IS_FILTER_STATUS_DONE, data };
}

export function setIsCreateTaskModalOpen(data) {
  return { type: SET_IS_CREATE_TASK_MODAL, data };
}
