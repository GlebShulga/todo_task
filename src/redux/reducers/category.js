import {
  ADD_CATEGORY,
  GET_CATEGORIES_LIST,
  GET_CATEGORIES_TITLE_LIST,
  GET_ROOT_CATEGORY_LIST,
  RENAME_CATEGORY,
  DEL_CATEGORY,
  UPDATE_CHOSEN_CATEGORY,
  SET_IS_FILTER_STATUS_DONE,
  SET_IS_CREATE_TASK_MODAL,
  SET_IS_OPEN_TASK_TABLE,
} from "../types/category";

const initialState = {
  categoryList: [],
  rootCategoryList: [],
  categoryTitleList: [],
  chosenCategory: {},
  isFilterStatusDone: false,
  isCreateTaskModalOpen: false,
  isOpenTaskTable: false,
};

// eslint-disable-next-line import/no-anonymous-default-export
export default (state = initialState, action) => {
  switch (action.type) {
    case GET_CATEGORIES_LIST:
    case ADD_CATEGORY:
    case DEL_CATEGORY:
    case RENAME_CATEGORY: {
      return { ...state, categoryList: action.categoryList };
    }
    case GET_ROOT_CATEGORY_LIST: {
      return { ...state, rootCategoryList: action.rootList };
    }
    case GET_CATEGORIES_TITLE_LIST: {
      return { ...state, categoryTitleList: action.titleList };
    }
    case UPDATE_CHOSEN_CATEGORY: {
      return { ...state, chosenCategory: action.chosenCategory };
    }
    case SET_IS_FILTER_STATUS_DONE: {
      return { ...state, isFilterStatusDone: action.data };
    }
    case SET_IS_CREATE_TASK_MODAL: {
      return { ...state, isCreateTaskModalOpen: action.data };
    }
    case SET_IS_OPEN_TASK_TABLE: {
      return { ...state, isOpenTaskTable: action.data };
    }
    default:
      return state;
  }
};


