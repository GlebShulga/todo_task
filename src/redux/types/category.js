const ADD_CATEGORY = "ADD_CATEGORY";
const GET_CATEGORIES_LIST = "GET_CATEGORIES_LIST";
const GET_CATEGORIES_TITLE_LIST = "GET_CATEGORIES_TITLE_LIST";
const GET_ROOT_CATEGORY_LIST = "GET_ROOT_CATEGORY_LIST";
const RENAME_CATEGORY = "RENAME_CATEGORY";
const DEL_CATEGORY = "DEL_CATEGORY";
const UPDATE_CHOSEN_CATEGORY = "UPDATE_CHOSEN_CATEGORY";
const SET_IS_FILTER_STATUS_DONE = "SET_IS_FILTER_STATUS_DONE";
const SET_IS_CREATE_TASK_MODAL = "SET_IS_CREATE_TASK_MODAL";
const SET_IS_OPEN_TASK_TABLE = "SET_IS_OPEN_TASK_TABLE";

module.exports = {
  ADD_CATEGORY,
  GET_CATEGORIES_LIST,
  GET_CATEGORIES_TITLE_LIST,
  GET_ROOT_CATEGORY_LIST,
  RENAME_CATEGORY,
  DEL_CATEGORY,
  UPDATE_CHOSEN_CATEGORY,
  SET_IS_FILTER_STATUS_DONE,
  SET_IS_CREATE_TASK_MODAL,
  SET_IS_OPEN_TASK_TABLE,
};
