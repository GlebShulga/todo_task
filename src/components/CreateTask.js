import React, { useState } from "react";
import axios from "axios";
import "../assets/scss/CreateTask.scss";

const CreateTask = ({ choosenCategory, setIsNewTaskCreated }) => {
  const [title, setTitle] = useState("");
  const [lengthError, setLengthError] = useState(false);
  const [noCategoryError, setNoCategoryError] = useState(false)

  const onChange = (e) => {
    setTitle(e.target.value);
    setLengthError(false);
  };

    const postTask = async () => {
      await axios({
        method: "post",
        url: "/api/v1/task",
        data: {
          title,
          categoryId,
        },
      });
    };

  const categoryId = choosenCategory.categoryId

  const onClickAddTask = async () => {
    if (categoryId) {
      if (title.length <= 15 && title.length >= 3) {
        await postTask()
        setIsNewTaskCreated(true);
      } else {
        setNoCategoryError(false);
        setLengthError(true);
      }
    } else {
      setNoCategoryError(true)
    }

  };
  const handleKeypress = (e) => {
    if (e.key === "Enter") {
      onClickAddTask();
    }
  };

  return (
    <div className="CreateTask">
      <div className="CreateTask-form">
        <input
          className="CreateTask-form_input"
          type="text"
          value={title}
          onChange={onChange}
          placeholder="Add new task"
          onKeyPress={handleKeypress}
        />
          <button
            type="button"
            className="CreateTask-form_button"
            onClick={onClickAddTask}
          >
            Add
          </button>
      </div>
      {lengthError && (
        <div className="Error">
          <div> The task length must not be shorter than 3 characters</div>
          <div>and</div>
          <div>must not exceed 15 characters.</div>
        </div>
      )}
      {noCategoryError && (
        <div className="Error">
          <div>Please, choose category</div>
        </div>
      )}
    </div>
  );
};

export default CreateTask;
