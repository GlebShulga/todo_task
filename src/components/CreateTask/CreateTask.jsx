import React, { useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import { addTask } from "../../redux/actions/task";
import "./CreateTask.scss";

const CreateTask = () => {
  const dispatch = useDispatch();
  const { chosenCategory } = useSelector((s) => s.category);
  const [title, setTitle] = useState("");
  const [lengthError, setLengthError] = useState(false);
  const [categoryError, setCategoryError] = useState(false);

  const onChange = (e) => {
    setTitle(e.target.value);
    setLengthError(false);
  };

  const categoryId = chosenCategory?.categoryId;

  const onClickAddTask = async () => {
    if (categoryId) {
      if (title?.trim().length <= 20 && title?.trim().length >= 3) {
        setCategoryError(false);
        setLengthError(false);
        dispatch(addTask(title, categoryId));
        setTitle("");
      } else {
        setLengthError(true);
        setCategoryError(false);
      }
    } else {
      setCategoryError(true);
    }
  };
  const handleKeypress = (e) => {
    if (e.key === "Enter") {
      onClickAddTask();
    }
  };

  return (
    <div className="create_task">
      <div className="create_task-form">
        <input
          className="create_task-form_input"
          type="text"
          value={title}
          onChange={onChange}
          placeholder="Add new task"
          onKeyPress={handleKeypress}
          data-testid="AddTaskInputField"
        />
        <button
          type="button"
          className="form_button_add"
          onClick={onClickAddTask}
          data-testid="AddTaskButton"
        >
          Add
        </button>
      </div>
      {lengthError && (
        <div className="error">
          <div> The task length must not be shorter than 3 characters</div>
          <div>and</div>
          <div>must not exceed 20 characters.</div>
        </div>
      )}
      {categoryError && (
        <div className="error">
          <div>Please, choose category</div>
        </div>
      )}
    </div>
  );
};

export default CreateTask;
