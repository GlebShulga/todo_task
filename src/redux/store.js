import { configureStore } from "@reduxjs/toolkit";
import category from "./reducers/category";
import task from "./reducers/task"

export const store = configureStore({
  reducer: {
    category,
    task
  },
});
