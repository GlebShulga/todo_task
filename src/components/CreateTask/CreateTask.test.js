import React from "react";
import { render, fireEvent } from "@testing-library/react";
import axios from "axios";
import store from "../../redux";
import { UPDATE_CHOSEN_CATEGORY } from "../../redux/types/category";
import CreateTask from "./CreateTask";
import AppTest from "../../../tests/AppTest";

jest.mock("axios");

describe("<CreateTask />", () => {
  it("should validate task category", () => {
    const { getByTestId, getByText } = render(
      <AppTest>
        <CreateTask />
      </AppTest>
    );

    fireEvent.click(getByTestId("AddTaskButton"));
    expect(getByText("Please, choose category")).toBeTruthy();
  });

  it("should validate task title minlength", () => {
    store.dispatch({
      type: UPDATE_CHOSEN_CATEGORY,
      chosenCategory: { categoryId: 1 },
    });

    const { getByTestId, getByText } = render(
      <AppTest>
        <CreateTask />
      </AppTest>
    );

    fireEvent.change(getByTestId("AddTaskInputField"), {
      target: { value: "" },
    });
    fireEvent.click(getByTestId("AddTaskButton"));

    const error = "The task length must not be shorter than 3 characters";
    expect(getByText(error)).toBeTruthy();
  });

  it("should add new task on click event", () => {
    axios.mockResolvedValue({});

    store.dispatch({
      type: UPDATE_CHOSEN_CATEGORY,
      chosenCategory: { categoryId: 1 },
    });

    const { getByTestId } = render(
      <AppTest>
        <CreateTask />
      </AppTest>
    );

    fireEvent.change(getByTestId("AddTaskInputField"), {
      target: { value: "This is a demo task" },
    });
    fireEvent.click(getByTestId("AddTaskButton"));

    expect(axios).toHaveBeenCalledWith({
      data: {
        categoryId: 1,
        title: "This is a demo task",
      },
      method: "post",
      url: "/api/v1/task",
    });
  });
});