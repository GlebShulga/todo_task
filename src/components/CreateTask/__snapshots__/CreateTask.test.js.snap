// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateTask Component CreateTask snapshot 1`] = `
<div
  className="create_task"
>
  <div
    className="create_task-form"
  >
    <input
      className="create_task-form_input"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Add new task"
      type="text"
      value=""
    />
    <button
      className="form_button_add"
      onClick={[Function]}
      type="button"
    >
      Add
    </button>
  </div>
</div>
`;

exports[`CreateTask Component CreateTask snapshot 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -18,6 +18,13 @@
        type=\\"button\\"
      >
        Add
      </button>
    </div>
+   <div
+     className=\\"error\\"
+   >
+     <div>
+       Please, choose category
+     </div>
+   </div>
  </div>"
`;
